File: habit-tracker/src/main/java/com/habit/habit_tracker/service/AuthService.java
package com.habit.habit_tracker.service;

import java.util.Optional;

import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.habit.habit_tracker.domain.User;
import com.habit.habit_tracker.dto.auth.LoginRequest;
import com.habit.habit_tracker.dto.auth.RegisterRequest;
import com.habit.habit_tracker.exception.ApiRequestException;
import com.habit.habit_tracker.repository.UserRepository;
import com.habit.habit_tracker.security.JwtProvider;

import static com.habit.habit_tracker.constants.ErrorMessage.*;

@Service
public class AuthService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtProvider jwtProvider;

    public AuthService(
            UserRepository userRepository,
            PasswordEncoder passwordEncoder,
            JwtProvider jwtProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtProvider = jwtProvider;
    }

    public String registerUser(RegisterRequest registerRequest) {
        Optional<User> existingUser = userRepository.findByUsername(registerRequest.getUsername());
        if (existingUser.isPresent()) {
            throw new ApiRequestException(USER_IN_USE, HttpStatus.CONFLICT);
        }

        User newUser = new User();
        newUser.setUsername(registerRequest.getUsername());
        newUser.setPassword(passwordEncoder.encode(registerRequest.getPassword()));

        userRepository.save(newUser);

        return jwtProvider.createToken(newUser.getId());
    }

    public String loginUser(LoginRequest loginRequest) {
        Optional<User> user = userRepository.findByUsername(loginRequest.getUsername());
        if (!user.isPresent()) {
            throw new ApiRequestException(USER_NOT_FOUND, HttpStatus.NOT_FOUND);
        }
        User foundUser = user.get();

        boolean matches = passwordEncoder.matches(loginRequest.getPassword(), foundUser.getPassword());
        if (!matches) {
            throw new ApiRequestException(INVALID_CREDENTIALS, HttpStatus.UNAUTHORIZED);
        }

        return jwtProvider.createToken(foundUser.getId());
    }
}

File: habit-tracker/src/main/kotlin/com/habit/habit_tracker/service/AuthService.kt
package com.habit.habit_tracker.service

import com.habit.habit_tracker.constants.ErrorMessage.USER_IN_USE
import com.habit.habit_tracker.constants.ErrorMessage.USER_NOT_FOUND
import com.habit.habit_tracker.constants.ErrorMessage.INVALID_CREDENTIALS

import com.habit.habit_tracker.domain.User
import com.habit.habit_tracker.dto.auth.AuthResult
import com.habit.habit_tracker.dto.auth.LoginRequest
import com.habit.habit_tracker.dto.auth.RegisterRequest
import com.habit.habit_tracker.exception.ApiRequestException
import com.habit.habit_tracker.repository.UserRepository
import com.habit.habit_tracker.security.JwtService
import com.habit.habit_tracker.security.UserPrincipal
import org.springframework.http.HttpStatus
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService
) {
    fun registerUser(request: RegisterRequest): AuthResult {
        if (userRepository.findByUsername(request.username).isPresent) {
            throw ApiRequestException(USER_IN_USE, HttpStatus.CONFLICT)
        }

        val newUser = User(
            username = request.username,
            password = passwordEncoder.encode(request.password)
        )

        userRepository.save(newUser)

        val userPrincipal = UserPrincipal(newUser)
        val token = jwtService.generateToken(userPrincipal)

        return AuthResult(token, newUser.username)
    }

    fun loginUser(request: LoginRequest): AuthResult {
        val user = userRepository.findByUsername(request.username)
            .orElseThrow { ApiRequestException(USER_NOT_FOUND, HttpStatus.NOT_FOUND) }

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw ApiRequestException(INVALID_CREDENTIALS, HttpStatus.UNAUTHORIZED)
        }

        val userPrincipal = UserPrincipal(user)
        val token = jwtService.generateToken(userPrincipal)


        return AuthResult(token, user.username)
    }
}
